### PR TITLE
Grid task logs filtering and local time

### PR DESCRIPTION
**Pull request by pierrejeambrun**
Follow up of https://github.com/apache/airflow/pull/24249.

Enforcing log parsing to add:
- Local datetime for the logs
- Filter on the log level (DEBUG, INFO, WARNING, etc.)
- Filter on the log group (python logger that emitted the log).

The parsing function is quite light at the moment, I still decided to memoized the results as I assume this could become quite expensive if the user requests the log `full_content`.

I also moved moment initialization for tests from testUtils to jest-setup.js. (Sometimes tests don't need a  wrapper, and we still have to import testUtils module to get moment defined)

Tested with mapped tasks and task group. ✔️ 

image
image

---
**^ Add meaningful description above**

Read the **[Pull Request Guidelines](https://github.com/apache/airflow/blob/main/CONTRIBUTING.rst#pull-request-guidelines)** for more information.
In case of fundamental code change, Airflow Improvement Proposal ([AIP](https://cwiki.apache.org/confluence/display/AIRFLOW/Airflow+Improvements+Proposals)) is needed.
In case of a new dependency, check compliance with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).
In case of backwards incompatible changes please leave a note in a newsfragement file, named `{pr_number}.significant.rst`, in [newsfragments](https://github.com/apache/airflow/tree/main/newsfragments).

### CODE REVIEW

File: airflow/www/static/js/grid/details/content/taskInstance/Logs/utils.js [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+/* global moment */
+
+import { defaultFormatWithTZ } from '../../../../../datetime_utils';
+
+export const LogLevel = {
```
Comment (bbovenzi) [inline airflow/www/static/js/grid/details/content/taskInstance/Logs/utils.js]:
```suggestion
export const logLevel = {
```
Generally, only classes or jsx components have a capitalized variable name

File: airflow/www/static/js/grid/details/content/taskInstance/Logs/index.jsx [hunk 2]
```diff
@@ -86,55 +94,95 @@ const Logs = ({
...
 
+  const { parsedLogs, logGroups = [] } = useMemo(() => parseLogs(
+    data,
+    timezone,
+    logLevelFilter,
+    logGroupFilter,
+  ),
+  [data, logGroupFilter, logLevelFilter, timezone]);
+
+  useEffect(() => {
```
Comment (bbovenzi) [inline airflow/www/static/js/grid/details/content/taskInstance/Logs/index.jsx:105]: Should we also reset the log level if it doesnt exist in the next task's logs?

File: airflow/www/static/js/grid/details/content/taskInstance/Logs/index.jsx [hunk 3]
```diff
@@ -65,6 +70,9 @@ const Logs = ({
   const [selectedAttempt, setSelectedAttempt] = useState(1);
   const [shouldRequestFullContent, setShouldRequestFullContent] = useState(false);
   const [wrap, setWrap] = useState(false);
+  const [logLevelFilter, setLogLevelFilter] = useState('');
+  const [logGroupFilter, setLogGroupFilter] = useState('');
```
Comment (bbovenzi) [inline airflow/www/static/js/grid/details/content/taskInstance/Logs/index.jsx]: Why did you go with calling it a log group? I'm trying to figure out if that is the best name or if we something like "file source"?

File: airflow/www/static/js/grid/details/content/taskInstance/Logs/index.jsx [hunk 4]
```diff
@@ -86,55 +94,95 @@ const Logs = ({
...
-                See More
-              </LinkButton>
-            </Flex>
+                {index}
+              </Button>
+            ))}
           </Flex>
-        </Box>
+          <Flex alignItems="center">
+            <Select
```
Comment (bbovenzi) [inline airflow/www/static/js/grid/details/content/taskInstance/Logs/index.jsx]: I think we'll want to use multi-select here. Although, I'm happy to do that in a subsequent PR to keep scope creep in check.

Comment (bbovenzi) [reply airflow/www/static/js/grid/details/content/taskInstance/Logs/index.jsx]: https://www.npmjs.com/package/chakra-react-select

File: airflow/www/static/js/grid/details/content/taskInstance/Logs/index.jsx [hunk 5]
```diff
@@ -86,55 +94,95 @@ const Logs = ({
     execution_date: executionDate,
   }).toString();
 
+  const { parsedLogs, logGroups = [] } = useMemo(() => parseLogs(
```
Comment (bbovenzi) [inline airflow/www/static/js/grid/details/content/taskInstance/Logs/index.jsx]: Good idea to memoize here! 👍

File: airflow/www/static/js/grid/details/content/taskInstance/Logs/index.jsx (hunk 2 above)
Comment (pierrejeambrun) [reply airflow/www/static/js/grid/details/content/taskInstance/Logs/index.jsx:105]: Normally there is no need for the logLevel as these options are hard coded in `utils/logLevel`. So there is no way options could "disappear" from one task to another. (Which is the case for the selected attempt number and source file filter if the new task do not have them).

File: airflow/www/static/js/grid/details/content/taskInstance/Logs/index.jsx (hunk 3 above)
Comment (pierrejeambrun) [reply airflow/www/static/js/grid/details/content/taskInstance/Logs/index.jsx]: Yes, we can rename that, updated

File: airflow/www/static/js/grid/details/content/taskInstance/Logs/index.jsx (hunk 4 above)
Comment (pierrejeambrun) [reply airflow/www/static/js/grid/details/content/taskInstance/Logs/index.jsx]: Nice idea :+1: 

Indeed maybe we should do that in its own PR so we can deliver the local datetime feature as fast as possible.

File: airflow/www/static/js/grid/details/content/taskInstance/Logs/index.jsx (hunk 3 above)
Comment (bbovenzi) [reply airflow/www/static/js/grid/details/content/taskInstance/Logs/index.jsx]: Yeah, let's go with that

File: airflow/www/static/js/grid/details/content/taskInstance/Logs/utils.js (hunk 1 above)
Comment (pierrejeambrun) [reply airflow/www/static/js/grid/details/content/taskInstance/Logs/utils.js]: :man_facepalming: , fixed.

**Comment by bbovenzi** [comment]
Let's at least set a `minWidth` to the file dropdown to the default text doesn't get cut off.
Let's set a fixed with to the log level dropdown, so it too doesn't move around.

Screen Shot 2022-06-13 at 4 27 39 PM
Screen Shot 2022-06-13 at 4 27 47 PM

File: airflow/www/static/js/grid/details/content/taskInstance/Logs/utils.test.js [hunk 6]
```diff
@@ -0,0 +1,110 @@
...
+      );
+
+      expect(fileSources).toHaveLength(expectedNumberOfFileSources);
+
+      const lines = parsedLogs.split('\n');
+      expect(lines).toHaveLength(expectedNumberOfLines);
+      lines.forEach((line) => expect(line).toContain(logLevelFilter));
+    });
+
+  test('parseLogs function with file source filter', () => {
```
Comment (bbovenzi) [inline airflow/www/static/js/grid/details/content/taskInstance/Logs/utils.test.js:91]: Let's also test parsing by log level.

Comment (pierrejeambrun) [reply airflow/www/static/js/grid/details/content/taskInstance/Logs/utils.test.js:91]: Just above  this one there is a test for filter on logLevel (INFO and WARNING),
https://github.com/apache/airflow/pull/24403/files/c409056433bc78d2cef3c9136e009cd84cf50c2d#diff-279abcb02ccc52bc014e3fe86d03e0f32b10ef59a0edc1c8a60fe564c94b91e4R70-R90

Comment (bbovenzi) [reply airflow/www/static/js/grid/details/content/taskInstance/Logs/utils.test.js:91]: Oh my bad. Yeah, lets do additional test for both. Why not?

Comment (pierrejeambrun) [reply airflow/www/static/js/grid/details/content/taskInstance/Logs/utils.test.js:91]: I added a test with both a filter on the logLevel and fileSource. (just for completeness)

https://github.com/apache/airflow/pull/24403/files#diff-279abcb02ccc52bc014e3fe86d03e0f32b10ef59a0edc1c8a60fe564c94b91e4R110-R127

**Review by bbovenzi — APPROVED** [review]
Great work! I think people will be really excited for all of this!

**Pull request merged**